### DISCUSSION TITLE
Custom trait doesn't persist after page reload

### DISCUSSION THREAD

**Discussion by dwr-ghostdev** [GrapesJS/grapesjs#5806] [Q&A]
I have a GrapesJs project that requires the video component to have a "muted" trait. It's part of the default attributes, but for some reason, the trait isn't there by default. So, I have to add it via a custom trait.

I'm having a problem with its persistence. The code works initially; by initially, after I drag the component onto the canvas - it shows the checkbox inside the Traits Manager and it also adds the "muted" attribute every toggle.

The problem is this: if I refresh the page or if I refresh the editor, the "muted" trait inside the Trait Manager disappears as well as the attributed set by trait in the element's attribute.

This is what I've tried so far:

```
export const customVideo = (editor) => {
  // Retrieve necessary components from DomComponents
  const domComps = editor.DomComponents;
  const dType = domComps.getType('video');
  const dModel = dType.model;
  const dView = dType.view;

  // Extend the default video model
  const customVideoModel = dModel.extend({
    // Custom initialization
    init() {
      dModel.prototype.init.apply(this, arguments);
      this.listenTo(this, 'change:provider', this.updateTraits);
      this.on('change:muted', this.onChangeMuted);
    },

    // Trigger the change event for the 'muted' attribute
    onChangeMuted() {
      const videoElement = this.view.el.children[0];

      if (!videoElement.getAttribute('muted')) {
        videoElement.setAttribute('muted', 'muted')
      } else {
        videoElement.removeAttribute('muted')
      }
    },

    // Update traits with the "muted" checkbox
    updateTraits() {
      const { em } = this;
      const prov = this.get('provider');
      let tagName = 'iframe';
      let traits;
  
      switch (prov) {
        case 'yt':
        case 'ytnc':
          traits = this.getYoutubeTraits();
          break;
        case 'vi':
          traits = this.getVimeoTraits();
          break;
        default:
          tagName = 'video';
          traits = [
            ...this.getSourceTraits(), 
            {
              type: 'checkbox',
              label: 'Muted',
              name: 'muted', // This corresponds to the attribute name
              changeProp: 1
            }
          ];
      }
  
      this.set({ tagName }, { silent: true }); // avoid break in view
      this.set({ traits });
      em.get('ready') && em.trigger('component:toggled');
    },
  });

  // Add the custom video component to GrapesJS
  domComps.addType('video', {
    model: customVideoModel,
    view: dView,

    // Update traits with the "muted" checkbox
    extend: 'video',
  });
} 
```

What seems to be the problem with this? Is there lacking code? 

Edit: added context and more details

Additional detail/s:
- I'm doing this for the "Source" provider ONLY.

For context:
This is what my Trait Manager looks like after I drag my Video component onto the canvas:
![image](https://github.com/GrapesJS/grapesjs/assets/147595564/c20bd45c-73e0-43ed-8a14-dbbbc35dc19c)

This is the video element after checking "Muted":
![image](https://github.com/GrapesJS/grapesjs/assets/147595564/7e0cc414-44cc-4dde-88e4-496725b0ce32)

After the page reload:
- Trait Manager:
- ![image](https://github.com/GrapesJS/grapesjs/assets/147595564/f2c66866-a05b-469b-9993-a87894223a6f)

- Video element (in the DOM)
- ![image](https://github.com/GrapesJS/grapesjs/assets/147595564/eda81f6a-5031-4b63-b8e6-7f4b4ec04770)

As you can see, it didn't have the "Muted" control and the "muted" attribute is missing as well.

**Comment by ctrhub** [comment]
Shouldn't the "muted" param be part of the video url? At least this is how it works on YouTube.
Another thing you might messing is that you need to update the view of the component on the first render with the component's settings(trait) using ```onRender() {}``` callback - https://arc.net/l/quote/lrrzfwze.
Example link - https://www.youtube.com/embed/lCiV4wACZ8w?autoplay=1&mute=1

**Reply by dwr-ghostdev** [reply]
Hi @ctrhub ! Thanks for responding! I forgot to mention that what I'm working on specifically is for the "Source" provider - which doesn't support query based parameters like you mentioned. It only gets muted once the "muted" attribute is added to the video element itself.

The code is working, actually. It adds the "muted" attribute already, but the problems are: 
- It doesn't bind to the element (but initially it adds since the "muted" control is present)
- The "mute" trait disappears after. 

Both are after a page reload.

**Reply by dwr-ghostdev** [reply]
Hi @ctrhub ! I updated the main question above. You may check it out for better context. Thanks!

**Reply by ctrhub** [reply]
It looks like the way to add this trait is a little different than what the documentation recommends.
https://grapesjs.com/docs/modules/Traits.html#customization
You only add your additional trait when something changes(and not on the first render). But it should be added to the list of traits of the model.
Something like this:
```JS
dModel.extend({
  defaults: {
      ...this.getSourceTraits(), 
      {
        type: 'checkbox',
        label: 'Muted',
        name: 'muted',
        changeProp: 1
      }
  },
  
  // ...
  init() {},
});
```

Or try to add this right after your ```updateTraits``` method:

```JS
onRender() {
  this.updateTraits();
}
```

**Reply by dwr-ghostdev** [reply]
I'll see what I can do with this approach. Thank you for your input, @ctrhub ! Appreciate it.

**Comment by dohomi** [comment]
My aim was to extend the video component with aspect ratio dropdown:
```ts
  const videoType = editor.DomComponents.getType("video")

  // Define the new trait
  const aspectRatioTrait = {
    type: "select",
    label: "Aspect Ratio",
    name: "aspectRatio",
    options: [
      { value: "16:9", name: "16:9" },
      { value: "4:3", name: "4:3" },
      { value: "1:1", name: "1:1" },
      { value: "21:9", name: "21:9" }
    ]
  }
  // Add the new trait to the video component's traits
  if (videoType) {
    const sourceTraits = videoType.model.prototype.getSourceTraits()
    const youtubeTraits = videoType.model.prototype.getYoutubeTraits()
    const vimeoTraits = videoType.model.prototype.getVimeoTraits()
    videoType.model.prototype.getSourceTraits = () => [aspectRatioTrait, ...sourceTraits]
    videoType.model.prototype.getYoutubeTraits = () => [aspectRatioTrait, ...youtubeTraits]
    videoType.model.prototype.getVimeoTraits = () => [aspectRatioTrait, ...vimeoTraits]

    editor.DomComponents.addType("video", videoType)
  }
```

**Reply by dwr-ghostdev** [reply]
Hi @dohomi ! Thanks for replying. Does your aspect ratio trait persist even after a page reload?

**Reply by dohomi** [reply]
yes the value persists through page reloads - if I persist the data correctly
